src/transformStateWithClones.js

```diff
@@ -7,7 +7,32 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let copyState = { ...state };
+  let newState = {};
+  const historyState = [];
+
+  actions.forEach((action) => {
+    switch (action.type) {
+      case 'addProperties':
+        newState = { ...copyState, ...action.extraData };
+        break;
+      case 'removeProperties':
+        newState = { ...copyState };
+        action.keysToRemove.forEach((key) => delete newState[key]);
+        break;
+      case 'clear':
+        newState = {};
+        break;
+      default:
+        newState = copyState;
+        break;
+    }
+
+    historyState.push(newState);
```

Comment:
Instead of pushing `newState` directly, you should push a copy of `newState` to ensure that subsequent modifications do not affect the already stored states in `historyState`. Use `{ ...newState }`.